lib/AST/SubstitutionMap.cpp

```diff
@@ -70,10 +70,19 @@ Type SubstitutionMap::lookupSubstitution(CanSubstitutableType type) const {
       genericEnv->mapTypeOutOfContext(archetype)->getCanonicalType());
   }
 
-  auto known = subMap.find(cast<GenericTypeParamType>(type));
+  auto genericParam = cast<GenericTypeParamType>(type);
+  auto known = subMap.find(genericParam);
   if (known != subMap.end() && known->second)
     return known->second;
 
+  // The generic parameter may have been made concrete by the generic signature,
```

Comment:
Why don't we just put the concrete type in the replacement map when we build the signature?